# Bài tập về nhà: Singular Value Decomposition (SVD)

Bài tập gồm 2 câu:
1. Phân rã một ma trận có rank thấp bằng SVD.
2. Chọn số thành phần k cho Truncated SVD dựa trên tỷ lệ năng lượng.

## Câu 1. Phân rã ma trận

Cho ma trận:
$$
A=
\begin{bmatrix}
1 & 0 & 1\\
0 & 1 & 1\\
1 & 0 & 1\\
0 & 1 & 1
\end{bmatrix}
\in\mathbb{R}^{4\times3}
$$

### 1. Chứng minh các cột của $A$ phụ thuộc tuyến tính và xác định $\operatorname{rank}(A)$.

In [1]:
import numpy as np

A = np.array([
    [1,0,1],
    [0,1,1],
    [1,0,1],
    [0,1,1]
])

print('Dễ thấy, cột 1 + cột 2 của ma trận A ra cột 3 nên các cột của A bị phụ thuộc tuyến tính')
print('\nRank của A:', np.linalg.matrix_rank(A))

Dễ thấy, cột 1 + cột 2 của ma trận A ra cột 3 nên các cột của A bị phụ thuộc tuyến tính

Rank của A: 2


### 2. Tính $A^\top A$.

In [3]:
AtA = A.T @ A
print(AtA)

[[2 0 2]
 [0 2 2]
 [2 2 4]]


### 3. Tìm các eigenvalues và eigenvectors trực chuẩn của $A^\top A$.

In [9]:
evalue, evector = np.linalg.eigh(AtA)

print('eigenvalue:', np.round(evalue, 4))
print('\neigenvector:\n', np.round(evector, 4))

eigenvalue: [0. 2. 6.]

eigenvector:
 [[-0.5774 -0.7071 -0.4082]
 [-0.5774  0.7071 -0.4082]
 [ 0.5774 -0.     -0.8165]]


### 4. Sắp xếp $\lambda_1\geq\lambda_2\geq\lambda_3$.

In [14]:
idx = np.argsort(evalue)[::-1]
evalue = evalue[idx]
evector = evector[:, idx]

print('Eigenvalue đã sắp xếp:', np.round(evalue, 4))

Eigenvalue đã sắp xếp: [6. 2. 0.]


### 5. Tính các singular values $\sigma_i=\sqrt{\lambda_i}$.

In [15]:
evalue[evalue < 0] = 0
singular_values = np.sqrt(evalue)
print('Singular values:', np.round(singular_values, 4))

Singular values: [2.4495 1.4142 0.    ]


### 6. Xác định các right singular vectors $v_1,v_2,v_3$.

In [18]:
print('\nRight singular vectors v1, v2, v3 theo thứ tự từ trái sang phải:\n', np.round(evector, 4))



Right singular vectors v1, v2, v3 theo thứ tự từ trái sang phải:
 [[-0.4082 -0.7071 -0.5774]
 [-0.4082  0.7071 -0.5774]
 [-0.8165 -0.      0.5774]]


### 7. Với mỗi $\sigma_i>0$, tính $u_i=Av_i/\sigma_i$.

In [20]:
U_partial = np.zeros((4,2))
for i in range(2): 
    U_partial[:,i] = (A @ evector[:, i]) / singular_values[i]

print('Các vector u1, u2:\n', np.round(U,4))

Các vector u1, u2:
 [[-0.5 -0.5]
 [-0.5  0.5]
 [-0.5 -0.5]
 [-0.5  0.5]]


### 8. Bổ sung các vector trực chuẩn còn thiếu để viết Full SVD: $A=U\Sigma V^\top$

In [21]:
U, S, Vt = np.linalg.svd(A, full_matrices = True)

print('U:\n', np.round(U, 4))
print('\nSigma:\n', np.round(S, 4))
print('\nV^T\n:', np.round(Vt,4))

U:
 [[-0.5  0.5  0.5 -0.5]
 [-0.5 -0.5 -0.5 -0.5]
 [-0.5  0.5 -0.5  0.5]
 [-0.5 -0.5  0.5  0.5]]

Sigma:
 [2.4495 1.4142 0.    ]

V^T
: [[-0.4082 -0.4082 -0.8165]
 [ 0.7071 -0.7071 -0.    ]
 [-0.5774 -0.5774  0.5774]]


### 9. Viết Reduced SVD: $A=U_r\Sigma_rV_r^\top$

In [26]:
k = 2
U_r = U[:,:k]
Sigma_r = np.diag(S[:k])
Vt_r = Vt[:k,:]

print('U_r:\n', np.round(U_r, 4))
print('\nSigma_r:\n', np.round(Sigma_r, 4))
print('\nV^T_r\n:', np.round(Vt_r,4))

U_r:
 [[-0.5  0.5]
 [-0.5 -0.5]
 [-0.5  0.5]
 [-0.5 -0.5]]

Sigma_r:
 [[2.4495 0.    ]
 [0.     1.4142]]

V^T_r
: [[-0.4082 -0.4082 -0.8165]
 [ 0.7071 -0.7071 -0.    ]]


### 10. Ghi rõ kích thước của $U_r,\Sigma_r,V_r^\top$.

In [29]:
print('Kích thước của U_r:', U_r.shape)
print('Kích thước của Sigma_r:', Sigma_r.shape)
print('Kích thước của V^T_r:', Vt_r.shape)

Kích thước của U_r: (4, 2)
Kích thước của Sigma_r: (2, 2)
Kích thước của V^T_r: (2, 3)


### 11. Kiểm tra $U^\top U=I$, $V^\top V=I$ và $U\Sigma V^\top=A$.

In [31]:
print('U^T . U == I:', np.allclose(U.T @ U, np.eye(4)))
print('V^T . V == I:', np.allclose(Vt @ Vt.T, np.eye(3)))

sigma_matrix = np.zeros((4,3))
np.fill_diagonal(sigma_matrix, S)
print('U.S.V^T == A:', np.allclose(U @ sigma_matrix @ Vt, A))

U^T . U == I: True
V^T . V == I: True
U.S.V^T == A: True


### 12. Giải thích:
- Vì sao có một singular value bằng 0?
- Singular vector tương ứng với singular value bằng 0 liên hệ thế nào với null space của $A$?
- Vì sao Reduced SVD vẫn tái tạo chính xác $A$?

- Ma trận A kích thước 4 x 3 nhưng các cột lại bị phụ thuộc tuyến tính, rank(A) = 2. Vì số lượng singular values khác 0 luôn bằng rank của ma trận nên singular value thứ 3 phải bằng 0
- Singular vector v3 chính là cơ sở cho null space của A. Dựa vào định nghĩa $Av_i = \sigma_i u_i$, nếu $\sigma_3 = 0$ thì $Av_3 = 0$, nghĩa là $v_3$ nằm trong null space của A.
- Vì $\sigma_3 = 0$, nên việc loại bỏ thành phần này không làm mất đi bất kỳ thông tin nào của A nên A vẫn được tái tạo chính xác hoàn toàn

## Câu 2. Chọn k cho Truncated SVD bằng tỷ lệ năng lượng

Một ma trận dữ liệu $X$ có các singular values:
$$
\sigma_1=12,\qquad
\sigma_2=5,\qquad
\sigma_3=3,\qquad
\sigma_4=1
$$

### 1. Xác định rank của $X$.

Vì hạng của ma trận luôn bằng số lượng singular values lớn hơn 0 nên rank(X) = 4

### 2. Tính tổng năng lượng: $E_{\text{total}}=\sum_{i=1}^{4}\sigma_i^2$

In [32]:
sigma = np.array([12,5,3,1])

E_total = np.sum(sigma ** 2)
print(E_total)

179


### 3. Với $k=1,2,3,4$, tính: $R(k)=\frac{\sum_{i=1}^{k}\sigma_i^2}{\sum_{i=1}^{4}\sigma_i^2}$

In [41]:
R_k = np.cumsum(sigma ** 2) / E_total

print(np.round(R_k,4))

[0.8045 0.9441 0.9944 1.    ]


### 4. Hoàn thành bảng:

| $k$ | Năng lượng giữ lại | Tỷ lệ $R(k)$ |
|---:|---:|---:|
| 1 | 144 | 0.8045  |
| 2 | 169 | 0.9441 |
| 3 | 178 | 0.9944 |
| 4 | 179 | 1 |

### 5. Tìm $k$ nhỏ nhất để giữ ít nhất:
- 80% năng lượng.
- 90% năng lượng.
- 95% năng lượng.
- 99% năng lượng.

Dựa vào kết quả của bảng trên, ta thấy:
- k = 1 cho ngưỡng 80%
- k = 2,3,4 cho ngưỡng 90%
- k = 3, 4 cho ngưỡng 95%
- k = 3, 4 cho ngưỡng 99%

### 6. Với $k=1,2,3$, tính: $\|X-X_k\|_F=\sqrt{\sum_{i=k+1}^{4}\sigma_i^2}$

In [43]:
error_f = []

for k in range(1,4):
    error = np.sqrt(np.sum(sigma[k:] ** 2))
    error_f.append(error)

print(np.round(error_f, 4))

[5.9161 3.1623 1.    ]


### 7. Hoàn thành bảng:

| $k$ | $R(k)$ | $\|X-X_k\|_F$ |
|---:|---:|---:|
| 1 | 0,8045 | 5.9161 |
| 2 | 0.9441 | 3.1623 |
| 3 | 0.9944 | 1 |

### 8. Giả sử $\sigma_4$ chủ yếu biểu diễn nhiễu. Chọn $k$ phù hợp và giải thích.

Trong trường hợp $\sigma_4$ chủ yếu biểu diễn nhiễu, ta chọn k = 3 để giữ lại với tỉ lệ 99.44% tổng năng lượng của dữ liệu gốc, vì thành phần này chiếm năng lượng rất nhỏ, việc cắt bỏ nó vừa giảm kích thước dữ liệu, vừa khử nhiễu làm cho dữ liệu xấp xỉ sạch hơn dữ liệu gốc

### 9. Trả lời:
- Vì sao singular values phải được sắp xếp giảm dần?
- Giữ càng nhiều singular values có phải lúc nào cũng tốt hơn không?
- Sự đánh đổi giữa tỷ lệ năng lượng và mức độ nén dữ liệu là gì?

- Vì khi SVD phân tách dữ liệu thì các thành phần mang nhiều năng lượng nhất hay nhiều thông tin nhất sẽ đứng đầu. Khi thực hiện Truncated SVD ta cũng bỏ đi các phần ít quan trọng nhất dễ dàng hơn
- Không phải giữ càng nhiều singular values lúc nào cũng tốt, nếu ma trận có kích thước quá lớn trong khi các singular values đó không đem lại giá trị đáng kể thì sẽ tốn bộ nhớ. Thêm nữa, các singular values nhỏ đứng cuối thường mang tính nhiễu, nếu ta giữ lại thì mô hình sẽ giữ các chi tiết quá cụ thể, dẫn đến tình trạng overfitting, cắt bỏ chúng có thể giúp mô hình học theo hướng khái quát hơn
- Sự đánh đổi giữa tỉ lệ năng lượng và mức độ nén dữ liệu là sự đánh đổi giữa chất lượng và kích thước, vì muốn tỉ lệ năng lượng cao, ta cần chọn k lớn, dẫn tới mức độ nén thấp, tốn nhiều bộ nhớ để lưu trữ và tính toán. Ngược lại, nếu muốn nén tốt, ta tiết kiệm được nhiều không gian nhưng phải chấp nhận việc mất mát thông tin.